src/runtime_src/core/common/api/xrt_kernel.cpp

```diff
@@ -1579,10 +1594,36 @@ class kernel_impl : public std::enable_shared_from_this<kernel_impl>
     m_usage_logger->log_kernel_info(device->core_device.get(), hwctx, name, args.size());
   }
 
-  // Delegating constructor with no module
   kernel_impl(std::shared_ptr<device_type> dev, xrt::hw_context ctx, const std::string& nm)
-    : kernel_impl{std::move(dev), std::move(ctx), {}, nm}
-  {}
+    : device(std::move(dev))                                            // share ownership
+    , hwctx(std::move(ctx))                                             // hw context
+    , hwqueue(hwctx)                                                    // hw queue
+    , m_module(get_module(hwctx, nm))                                   // module object with matching kernel name
+    , properties(xrt_core::module_int::get_kernel_info(m_module).props) // kernel info present in Elf
+    , uid(create_uid())
+  {
+    XRT_DEBUGF("kernel_impl::kernel_impl(%d)\n", uid);
+
+    // initialize kernel name and ctrl code index
+    auto i = nm.find(":");
+    if (i == std::string::npos) {
+      // default case - ctrl code 0 will be used
+      name = nm.substr(0, nm.size());
+      m_ctrl_code_index = 0;
+    }
+    else {
+      name = nm.substr(0, i);
+      m_ctrl_code_index = std::stoul(nm.substr(i+1, nm.size()-i-1));
+    }
```

Comment:
Uhh, this is a little hard to parse.
This is the second place name is parsed at `:`.  Can this parsing be re-factored into say
```
std::pair<std::string, uint32_t>
static get_kernel_name_and_ctrl_code_index(const std::string& nm)
{
   ...
   return {name, index};
}
```
Then use as 
```
std::tie(name, m_ctrl_code_index) = get_name_and_ctrl_code_index(nm);
```
Maybe choose a better function name for the refactored code.

Reply:
I will refactor the code Soren, my attempt was to use initializer list for most of the class members but yeah it does not look good.